engines/ic3.cpp

```diff
@@ -0,0 +1,299 @@
+  }
+
+  UnorderedTermSet keep, core_set;
+  TermVec bool_assump, tmp, new_tmp, removed, lits;
+  lits = c.children;
+
+  if (options_.random_seed_ > 0) {
+    shuffle(
+        lits.begin(), lits.end(), default_random_engine(options_.random_seed_));
+  }
+
+  int iter = 0;
+  bool progress = true;
+  while (iter <= options_.ic3_gen_max_iter_ && lits.size() > 1 && progress) {
+    iter = options_.ic3_gen_max_iter_ > 0 ? iter + 1 : iter;
+    size_t prev_size = lits.size();
+    for (auto a : lits) {
+      // check if we can drop a
+      if (keep.find(a) != keep.end()) {
+        continue;
+      }
+      tmp.clear();
+      for (auto aa : lits) {
+        if (a != aa) {
+          tmp.push_back(aa);
+        }
+      }
+
+      Term tmp_and_term = make_and(tmp);
+      if (!intersects_initial(tmp_and_term)) {
+        assert(solver_context_ == 0);
+        push_solver_context();
+        assert_frame_labels(i - 1);
+        assert_trans_label();
+        solver_->assert_formula(solver_->make_term(Not, tmp_and_term));
+
+        Term l;
+        bool_assump.clear();
+        for (auto t : tmp) {
+          l = label(t);
+          solver_->assert_formula(solver_->make_term(Implies, l, ts_->next(t)));
+          bool_assump.push_back(l);
+        }
+
+        Result r = solver_->check_sat_assuming(bool_assump);
+        assert(!r.is_unknown());
+
+        if (r.is_sat()) {
+          // we cannot drop a
+          pop_solver_context();
+        } else {
+          new_tmp.clear();
+          removed.clear();
+          core_set.clear();
+          // filter using unsatcore
+          solver_->get_unsat_core(core_set);
+          for (size_t j = 0; j < bool_assump.size(); ++j) {
+            if (core_set.find(bool_assump[j]) != core_set.end()) {
+              new_tmp.push_back(tmp[j]);
+            } else {
+              removed.push_back(tmp[j]);
+            }
+          }
+
+          pop_solver_context();
+
+          // keep in mind that you cannot drop a literal if it causes c to
+          // intersect with the initial states
+          size_t size = new_tmp.size();
+          fix_if_intersects_initial(new_tmp, removed);
+          // remember the literals which cannot be dropped
+          for (size_t i = size; i < new_tmp.size(); ++i) {
+            keep.insert(new_tmp[i]);
+          }
+
+          lits = new_tmp;
+          break;  // next iteration
+        }
+      }
+    }
+
+    progress = lits.size() < prev_size;
+  }
+
+  // TODO: would it be more intuitive to start with a clause
+  //       and generalize the clause directly?
+  IC3Formula blocking_clause =
+      ic3_formula_negate(ic3_formula_conjunction(lits));
+  assert(blocking_clause.is_disjunction());  // expecting a clause
+  return { blocking_clause };
+}
+
+IC3Formula IC3::generalize_predecessor(size_t i, const IC3Formula & c)
+{
+  // TODO: change this so we don't have to depend on the solver context to be
```

Comment:
My point here is it would be nice if we didn't have these assumptions about the solver state. I didn't want someone implementing the derived class to have to pop the solver context, so that's done after calling this method instead of in it. But that also means that you can't use the solver in `generalize_predecessor` because it's still at context level 1 and has assertions from the relative inductiveness check.